### ISSUE TITLE
incorrect use of export modifiers

### ISSUE THREAD

**Issue by pps83** [libgd/libgd#286]
Since gd.h declares API functions using BGD_DECLARE, then it also has to use that same modifier in .c files as well. ~~I think some paper by Ulrich Drepper~~ [Step-by-step_guide](https://gcc.gnu.org/wiki/Visibility#Step-by-step_guide) describes how to do that properly. In short, if it's used on declarations in header then it also has to be inside .c files. Because it's not done properly windows build fails:

```
gd_webp.c:19:12: error: conflicting types for 'gdImageCreateFromWebp'
 gdImagePtr gdImageCreateFromWebp (FILE * inFile)
            ^~~~~~~~~~~~~~~~~~~~~
```

BGD_DECLARE is defined as follows on windows:
`#define BGD_DECLARE(rt) BGD_EXPORT_DATA_PROT rt BGD_STDCALL`

that weird `stdcall` IMO should be removed and because of that `stdcall` declarations inside .c files are now conflicting, because they are compiled `cdecl`.
gd.h says that "2.0.24: __stdcall also needed for Visual BASIC and other languages". Not sure about VB and others, but I find it hard to believe that VB and other languages don't support normal c calling conventions.

**Comment by vapier** [comment]
no, you don't need to duplicate markings in .c files.  you just mark the prototype (i.e. gd.h) and include it in the .c files.

please describe your actual build environment.

**Comment by pps83** [comment]
That error message comes from gcc. I took libgd as-is and tried to compile it using mingw32 on msys2. Nothing special. Check that Step-by-step_guide link from gcc that describes how to do that properly. Regarding "you don't need to duplicate markings in .c files" - in some cases you are ok without it, but in others it won't be correct. For example, when you build your code file that compiles that .c doesn't need to include its corresponding .h where the function is declared. Then, if you don't duplicate it in .c you end up with this code:
```

// .h
__declspec(dllexport) void __stdcall sumeFunc(a, b, c);
...
// .c
void sumeFunc(a, b, c)
{
 ... test
}
```

how possibly these two can be identical, just because they have identical names? That's why gcc produces error there.

**Comment by vapier** [comment]
i'm aware that if you don't include the header file, the marking propagation is missing.  but that isn't relevant to gd as we include the header everywhere.  in fact, the only way your error shows up is because we made sure to include gd.h before we defined the function.

i've built for mingw as well and it works fine for me.  please post the full commands you're running as well as the full output.  you can compress & attach logs to github.

**Comment by cmb69** [comment]
It appears to me that Pavel's issue has already been solved with commit 5cd8336c, which is also in GD-2.2.

**Comment by pps83** [comment]
@cmb69 yes, looks like it would fix the issue. However, `BGD_DECLARE(void) functionName(...)` could have been made in a more traditional way, something like `BGD_API void functionName(...)`

**Comment by cmb69** [comment]
@pps83 A version with this commit has not yet been released (it will probably be contained in gd-2.2.4, though). You can checkout the master or GD-2.2 branch, if you want to test it.

Wrt. BDG_DECLARE(): that seems to be uncommon, but it had been [introduced with gd-2.0.24](https://github.com/libgd/libgd/blob/GD_2_0_24/src/gd.h#L14-L23), so we're bound to it (at least for now). If you think it should be changed for gd-3, please open another issue.
